add: mecha cage, mecha alerts

NuSanya opened the pull request:
## Описание

почему то был мёрж конфликт, не залили. предложка баунти, добавляет новый модуль и алёрты на мех

## Причина создания ПР / Почему это хорошо для игры

https://discord.com/channels/617003227182792704/755125334097133628/1284950351673430187

## Тесты

https://www.youtube.com/watch?v=nMl4fU3tho4
я не проверял сейчас, видео довольно старое

Honker9999999 committed:
add: mecha cage

NuSanya committed:
Update goon_vampire_powers.dm

NuSanya committed:
Update goon_vampire_powers.dm

NuSanya commented:
мёрж конфликт не из-за моего кода, если правильно понимаю. при убирании кода из code/modules/antagonists/goon_vampire/goon_vampire_powers.dm остаётся конфликт.

NuSanya committed:
Update goon_vampire_powers.dm

NuSanya committed:
check

NuSanya committed:
Merge branch 'master220' into cage

NuSanya committed:
Create goon_vampire_powers.dm

NuSanya committed:
Delete goon_vampire_powers.dm

NuSanya committed:
Update goon_vampire_powers.dm

Antoonij commented on code/_onclick/hud/alert.dm:
```diff
@@ -563,6 +563,38 @@ so as to remain in compliance with the most up-to-date laws."
+
+//MECH MODULES
+
+//cage module
```
```suggestion
// MECH MODULES

// cage module
```
Codestyle

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
+		return TRUE
+
+	occupant_message(span_notice("[target] can't be suppressed, since [target] is not in critical condition"))
+	return FALSE
```
такое добро нельзя держать в 1 проке, нужно разделить по прокам

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
+	var/mob/living/carbon/holding
+	var/turf/holding_turf
+	var/datum/action/innate/mecha/select_module/button // for custom icons
+	var/current_alert //wacky case
```
1. Тут лучше раписать автодок
2. Автодок начинается с 3 слешей и находится над переменной

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
 		W.fast_pressure_step_in = initial(W.fast_pressure_step_in)
+
+
+//SCS-3 CAGE
```
```suggestion
// SCS-3 CAGE
```
codestyle

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
+/obj/item/mecha_parts/mecha_equipment/cage/attach_act(obj/mecha/M)
+	START_PROCESSING(SSobj, src)
+
+/obj/item/mecha_parts/mecha_equipment/cage/Destroy()
+	for(var/atom/movable/AM in src)
+		AM.forceMove(get_turf(src))
+		if(holding)
+			stop_supressing(holding)
```
prisoner, holding, holding turf должны в этом проке приводиться к нуллу

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
+	button.button_icon_state = icon
+	button.UpdateButtonIcon()
+
+/obj/item/mecha_parts/mecha_equipment/cage/proc/change_alert(var/stage_number) //zero, one, two, three in string
+	var/mob/living/carbon/H = chassis.occupant
+	switch(stage_number)
+		if("zero")
+			H.throw_alert(alert_category, /atom/movable/screen/alert/mech_cage/zero)
+		if("one")
+			H.throw_alert(alert_category, /atom/movable/screen/alert/mech_cage/one)
+		if("two")
+			H.throw_alert(alert_category, /atom/movable/screen/alert/mech_cage/two)
+		if("three")
+			H.throw_alert(alert_category, /atom/movable/screen/alert/mech_cage/three)
```
В  твоих алертах можно и хранить стадию в переменной (и использовать дефайны, а не просто текст) и обращаться к ним через for (Т.е subtypesof на /atom/movable/screen/alert/mech_cage), находя нужный алерт по переданной стадии в проке и тровать его, записывать стадию в переменную current_alert (комментарием ниже прошу сменить название).

Antoonij replied:
Впрочем лучше сделай ассоциативный список, пожертвуем удобством

NuSanya replied:
сделал в итоге через subtypes

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,280 @@
+		if("three")
+			H.throw_alert(alert_category, /atom/movable/screen/alert/mech_cage/three)
+
+	current_alert = stage_number
```
название переменной current_alert не соответствует своему названию, а содержит в себе стадию. Изменить ее название.

NuSanya committed:
codestyle 1

Co-authored-by: Antoonij <42318445+Antoonij@users.noreply.github.com>

NuSanya committed:
codestyle 2

Co-authored-by: Antoonij <42318445+Antoonij@users.noreply.github.com>

NuSanya committed:
cleaning up code

NuSanya committed:
Merge branch 'cage' of https://github.com/NuSanya/Paradise into cage

NuSanya committed:
oops

Antoonij commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,294 @@
+	if(do_after(prisoner, 30 SECONDS, prisoner))
+		eject(TRUE)
+
+/obj/item/mecha_parts/mecha_equipment/cage/process()
+	if(holding || prisoner)
+		if(holding)
+			var/turf/actual_turf = get_turf(holding)
+			if(actual_turf != holding_turf)
+				stop_supressing(holding)
+				actual_turf = null
+			if(chassis.occupant == null)
+				stop_supressing(holding)
+
+		if(prisoner)
+			if(!istype(prisoner.loc, src))
+				prisoner = null
+		else if(current_alert == "three")
+			if(holding.handcuffed)
+				change_alert("two")
+			else
+				change_alert("one")
+		else if(button)
+			if(button.button_icon_state == "mecha_cage_activated")
+				change_state("mecha_cage")
+
+	else
+		if(chassis.occupant)
+			if(current_alert != "zero" && chassis.selected == src)
+				change_alert("zero")
+		if(button)
+			if(button.button_icon_state == "mecha_cage_activated")
+				change_state("mecha_cage")
```
В данном PRе имеется огромное количество проблем, пока что в таком виде это не попадет в игру.

Передвижение заключенного можно отслеживать через сигнал COMSIG_MOVABLE_MOVED, а не каждый тик в процессе

Хендкаффы также можно отслеживать через сигнал (правда новый придется сделать) и менять алерт в зависимости от того, анкаффнулся ли заключенный или нет. Это также не нужно проверять каждый тик в процессинге.

NuSanya replied:
Тяжело

Antoonij replied:
Посмотри как работает RegisterSignal и SEND_SIGNAL. 

Нужно будет зарегистрировать сигнал COMSIG_MOVABLE_MOVED на заключенном в моменте, когда заключенный присваивается, в сигнальном проке уже проверять как куда он двинулся.

Сначала потренируйся на этом, потом перейдешь к отправке сигналов на анкафф/хедкафф

NuSanya committed:
alert defines, change_alert proc remake

NuSanya committed:
tweaks

removes processing, fixes some bugs, adds some sprites that got deleted when this pr got copied by another person

NuSanya committed:
Update paradise.dme

NuSanya committed:
Merge branch 'master220' into cage

PlayerUnknown14 reviewed (changes requested):
Запрещаю добавление в игру, пока весь добавляемый контент не будет предварительно русифицирован.

NuSanya committed:
перевод 1

NuSanya committed:
перевод 2

NuSanya committed:
перевод 3

NuSanya committed:
перевод 4

NuSanya committed:
перевод 5

NuSanya committed:
перевод 6

NuSanya committed:
перевод 7

NuSanya commented:
переведено

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+// SCS-3 CAGE
+
+/obj/item/mecha_parts/mecha_equipment/cage
+	name = "Клетка SCS-3"
```
Названия не нужно переводить, пусть будет на английском.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/item/mecha_parts/mecha_equipment/cage
+	name = "Клетка SCS-3"
+	desc = "Модуль для экзокостюмов, используемый в задержании преступников."
```
```suggestion
	desc = "Модуль для экзокостюмов, используемый для задержании преступников."
	ru_names = list(
	    NOMINATIVE = "модуль \"Клетка SCS-3\"",
	    GENITIVE = "модуля \"Клетка SCS-3\"", 
	    DATIVE = "модулю \"Клетка SCS-3\"", 
	    ACCUSATIVE = "модуль \"Клетка SCS-3\"", 
	    INSTRUMENTAL = "модулем \"Клетка SCS-3\"", 
	    PREPOSITIONAL = "модулю \"Клетка SCS-3\""
	)
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+		insert_action(target)
+		return TRUE
+
+	occupant_message(span_notice("[target] не может быть удержа[genderize_ru(target.gender, "н", "на", "но", "ны")], так как [target] не находится в критическом состоянии"))
```
```suggestion
	occupant_message(span_notice("[target] не может быть удержа[genderize_ru(target.gender, "н", "на", "но", "ны")], так как [target] не наход[pluralize_ru(targer.gender, "ит", "ят")]ся в критическом состоянии."))
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/item/mecha_parts/mecha_equipment/cage/proc/supress_action(mob/living/carbon/target)
+	if(holding)
+		occupant_message(span_notice("Ты перестаёшь удерживать [holding], и начинаешь удерживать [target]..."))
```
```suggestion
		occupant_message(span_notice("Вы перестаёте удерживать [holding] и начинаете удерживать [target]..."))
```
Ты -> Вы здесь и ниже.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+/obj/item/mecha_parts/mecha_equipment/cage/proc/supress_action(mob/living/carbon/target)
+	if(holding)
+		occupant_message(span_notice("Ты перестаёшь удерживать [holding], и начинаешь удерживать [target]..."))
+		chassis.visible_message(span_warning("[chassis] перестаёт удерживать [holding] и сменяется на [target]."))
```
```suggestion
		chassis.visible_message(span_warning("[capitalize(chassis.declent_ru(NOMINATIVE))] перестаёт удерживать [holding] и начинает удерживать [target]."))
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+	UnregisterSignal(target, COMSIG_MOVABLE_MOVED)
+	RegisterSignal(target, COMSIG_MOVABLE_MOVED, PROC_REF(on_escape))
+	update_equip_info()
+	occupant_message(span_notice("[target] успешно загруже[genderize_ru(target.gender, "н", "на", "но", "ны")]."))
```
```suggestion
	occupant_message(span_notice("[target] успешно помещ[genderize_ru(target.gender, "ён", "ена", "ено", "ены")] в клетку."))
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+/obj/item/mecha_parts/mecha_equipment/cage/proc/on_escape(mob/living/carbon/target)
+	SIGNAL_HANDLER
+	occupant_message("[prisoner] сбежа[genderize_ru(target.gender, "л", "ла", "ло", "лы")].")
+	log_message("[prisoner] сбежа[genderize_ru(target.gender, "л", "ла", "ло", "лы")].")
```
Логи не переводим, они на английском.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/item/mecha_parts/mecha_equipment/cage/proc/on_escape(mob/living/carbon/target)
+	SIGNAL_HANDLER
+	occupant_message("[prisoner] сбежа[genderize_ru(target.gender, "л", "ла", "ло", "лы")].")
```
Пропиши тут `span_warning` или типа того, чтобы текст выделялся.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/item/mecha_parts/mecha_equipment/cage/proc/prisoner_insertion_check(mob/living/carbon/target)
+	if(target.buckled)
+		occupant_message(span_warning("[target] не помест[pluralize_ru(target.gender, "ится", "ятся")] в клетку, так как [target] прикова[genderize_ru(target.gender, "н", "нна", "нно", "нны")] к [target.buckled]!"))
```
```suggestion
		occupant_message(span_warning("[target] не помест[pluralize_ru(target.gender, "ит", "ят")]ся в клетку, так как [target] прикова[genderize_ru(target.gender, "н", "на", "но", "ны")] к [target.buckled.declent_ru(DATIVE)]!"))
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+		occupant_message(span_warning("[target] не помест[pluralize_ru(target.gender, "ится", "ятся")] в клетку, так как [target] прикова[genderize_ru(target.gender, "н", "нна", "нно", "нны")] к [target.buckled]!"))
+		return FALSE
+	if(target.has_buckled_mobs())
+		occupant_message(span_warning("[target] не помест[pluralize_ru(target.gender, "ится", "ятся")] в клетку из-за прикованных к [genderize_ru(target.gender, "нему", "ней", "нему", "ним")] животным!"))
```
`has_buckled_mobs` относится только к слаймам, поэтому тут мы конкретно прописываем, что речь про них.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/item/mecha_parts/mecha_equipment/cage/can_detach()
+	if(prisoner || holding)
+		occupant_message(span_warning("Невозможно отсоединить [src] - модуль работает!"))
```
```suggestion
		occupant_message(span_warning("Невозможно отсоединить [declent_ru(ACCUSATIVE)] - модуль работает!"))
```

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,285 @@
+
+/obj/effect/supress
+	name = "Клешни экзокостюма"
+	desc = "Сейчас кого-то своруют..."
```
Аналогично выше.

PlayerUnknown14 commented on code/game/mecha/mecha.dm:
```diff
@@ -409,6 +410,11 @@
 			occupant_message(span_danger("Unable to move while in zoom mode."))
```
Здесь и выше перевести.

PlayerUnknown14 reviewed (changes requested):
Буквы - доработать.

NuSanya committed:
blyat1

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyat2

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyat3

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyat4

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyatiazabylkakoi

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyat666

Co-authored-by: PlayerUnknown14 <139672964+PlayerUnknown14@users.noreply.github.com>

NuSanya committed:
blyat1337

NuSanya committed:
test

NuSanya committed:
fix

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,293 @@
+		supress(target)
+	else
+		occupant_message(span_notice("Вы начинаете удерживать [target]..."))
+		chassis.visible_message(span_warning("[chassis] начинает удерживать [target]."))
```
Здесь и ниже прописать декленты, аналогично сообщению на 688 строке.

PlayerUnknown14 commented on code/game/mecha/equipment/tools/other_tools.dm:
```diff
@@ -590,3 +590,293 @@
+
+/obj/effect/supress
+	name = "Mech claws"
+	desc = "Сейчас кого-то своруют..."
```
Ру неймы.

PlayerUnknown14 commented on code/game/mecha/mecha.dm:
```diff
@@ -409,6 +410,11 @@
+	if(locate(/obj/item/mecha_parts/mecha_equipment/cage) in equipment)
+		var/obj/item/mecha_parts/mecha_equipment/cage/H = locate(/obj/item/mecha_parts/mecha_equipment/cage) in equipment
+		if(H.holding)
+			occupant_message(span_notice("Ты перестаёшь удерживать [H.holding]."))
```
"Ты" нигде не прописываем в сообщениях.

PlayerUnknown14 commented on code/modules/antagonists/changeling/powers/biodegrade.dm:
```diff
@@ -49,6 +49,21 @@
+	if(istype(user.loc, /obj/item/mecha_parts/mecha_equipment/cage))
+		var/obj/item/mecha_parts/mecha_equipment/cage/container = user.loc
+		var/obj/mecha/mech = container.chassis
+		mech.visible_message(span_warning("Камера содержания [mech] начинает плавиться!"), \
```
Декленты

PlayerUnknown14 reviewed (changes requested):
Почти.

NuSanya committed:
у меня 69 по егэ по русскому

NuSanya committed:
Update mecha_equipment.dmi

NuSanya committed:
test

NuSanya committed:
test2

NuSanya commented:
сделаю нормальный форк и пересоздам
